ide/coqide/idetop.ml

```diff
@@ -367,6 +367,26 @@ let export_option_state s = {
   Interface.opt_value = export_option_value s.Goptions.opt_value;
 }
 
+exception NotSupported of string
+
+let top_cmd diff_type =
+  let open G_toplevel in
+  let pa = Pcoq.Parsable.make (Stream.of_string diff_type) in
+  let doc = Stm.get_doc 0 in
+  let state = Stm.get_current_state ~doc in
+  begin match Coqloop.read_sentence_ide doc state pa with
+    | Some G_toplevel.VernacDrop
+    | None -> ()
+    | Some stm -> begin
+      match stm with
+      | VernacShowGoal _
+      | VernacShowProofDiffs _ ->
+  (*    | todo: add "Show" command *)
```

Comment:
I'm looking at the code to propose some changes, but here I need to know a bit more about the intended use of these commands. To my understanding you added a menu item to request the display of a diff, and this is the role played by `VernacShowProofDiffs`. What about `VernacShowGoal` (which is for now doing the same)? And what about `Show`?
(There is already a message, called goals, which gets the goals in order to print them)

Reply:
Does ppx have an option to work with XML?  I didn't see mention of that in a quick scan of their readme file.  Or would you consider changing the protocol to use JSON?

> To my understanding you added a menu item to request the display of a diff, and this is the role played by `VernacShowProofDiffs`.

Exactly.

`Show Goal` is used by the Prooftree tool developed by @hendriktews.  See https://github.com/coq/coq/issues/10315.  I think a menu item is not needed (and easily added later).  You'll return it as Feedback (minimal change on the PG side) or as structured data?

`Show` may not be necessary for PG if `Goals` is equivalent.  Hopefully not requiring too much change in PG.  It was an issue for the `coqtop`-based protocol that we worked around.

Reply:
I'd rather return a structured data for ShowGoal. Maybe it can just be an extra piece of info for the Goals message: today it gets the goals at the tip of the document, while ShowGoal seems to let you chose the state id, so it is not very different (it's as if the default state id was the tip of the document).

For ShowProofDiffs I plan to return something more structured that can be printed client side. Maybe the richpp data type which already has XML serialization (colors are tags around text, not vt100 codes, and the client can decide which color to use for each tag). But I did not looked at this yet.

Wrt ppx, one could, maybe should, generate them all. S-expressions would work fine for emacs, json for other clients... It's the same syntax tree, it's only the "form of the parentheses" that would change. The "main doc" would be the OCaml data type.

Reply:
> I'd rather return a structured data for ShowGoal.

And keep the `Show Goal` command?  Hopefully the two code paths won't diverge.  Using the command in coqtop is good for developers to experiment with the behavior and for use in the test suite.  But we don't have a good unit-test solution for the protocol: the test suite isn't set up to verify return values of the protocol calls in `fake_ide`, only AFAICT whether the protocol fails.  We should have that capability and some automated tests, especially if we expect further protocol enhancements.

> Maybe the richpp data type which already has XML serialization

Yes, reuse that.  Feedback and Goals both use that.

> Wrt ppx, one could, maybe should, generate them all. S-expressions would work fine for emacs, json for other clients... 

PPX might let the client pick which format to use: If the first message begins with `<`, use XML; if it begins with `{`, use JSON; if it begins with `(`, use S-expressions.

Reply:
Yes, I'll keep the coqtop commands as they are. I think I will split their implementation so that the most of the code is shared, and separate the code path for tty and xml. At least this is my plan.

I totally agree about the testing part in general. For fake_ide I believe it is doable to implement an assertion on the value returned by a command, while it is hard to assert about "casual" feedback (the link between a feedback and the command that triggers it is not always obvious). Let's see how it goes.